modules/replication/module.py

```diff
@@ -55,18 +61,162 @@ def __init__(self, id, resolver, n, f, k):
+                processor_ids = []
+                for i, replica_structure in self.rep:
+                    if replica_structure[VIEW_CHANGE] and self.prim == i:
+                        processor_ids.add(i)
+
+                if len(processor_ids) > (3 * self.number_of_byzantine) + 1:
+                    self.renew_reqs(processor_ids)
+                    self.find_cons_state(self.com_pref_states(
+                        (3 * self.number_of_byzantine) + 1
+                    ))  # TODO assign return val from cons_state when impl.
+                    self.rep[self.id][VIEW_CHANGE] = False
+            # lines 7-8
+            elif self.rep[self.id][VIEW_CHANGE] and \
+                    (self.rep[self.prim][VIEW_CHANGE] is False and
+                     (self.rep[self.id][PRIM] == self.rep[self.prim][PRIM])):
+                processor_ids = []
+                for i in range(self.number_of_nodes):
+                    if self.resolver.execute(
+                        Module.VIEW_ESTABLISHMENT_MODULE,
+                            Function.GET_CURRENT_VIEW, i) \
+                            == self.prim:
+                        processor_ids.append(i)
+                if len(processor_ids) >= (4 * self.number_of_byzantine) + 1 \
+                   and self.check_new_v_state(self.prim):
+                    self.rep[self.id] = self.rep[self.prim]
+                    self.rep[self.id][VIEW_CHANGE] = False
+
+            # lines 9 - 10
+            X = self.find_cons_state(self.com_pref_states(
+                (3 * self.number_of_byzantine) + 1
+            ))
+            Y = self.get_ds_state()
+            if len(X) == 0 and len(Y) > 0:
+                X = Y
+
+            # lines 11 - 14
+            self.rep[self.id][CON_FLAG] = (len(X) > 0)
+            if not (self.rep[self.id][CON_FLAG]) and \
+               (not (self.is_prefix_of(self.rep[self.id][REP_STATE], X)) or
+               self.rep[self.id][REP_STATE] == self.DEF_STATE[REP_STATE] or
+                    self.delayed()):
+                self.rep[i][REP_STATE] = X
+            if self.stale_rep() or self.conflict():
+                self.flush_local()
+                self.rep[self.id]
+                self.need_flush = True
+            if self.flush:
+                self.flush_local()
+
+            self.rep[self.id][PEND_REQS].append(self.known_pend_reqs())
+
+            # line 15 - 25
+            if (self.resolver.execute(
+                    Module.VIEW_ESTABLISHMENT_MODULE,
+                    Function.ALLOW_SERVICE) and (self.need_flush is False)):
+                if self.resolver.execute(
+                    Module.PRIMARY_MONITORING_MODULE,
+                    Function.NO_VIEW_CHANGE) \
+                        and self.rep[self.id][VIEW_CHANGE] is False:
+                    if self.id == self.prim:
+                        # primary processes all pending reqs
+                        for req in self.get_pend_reqs():
```

Comment:
code uses `self.rep[self.id][PEND_REQS]`, the get_pend_reqs() is used by Primary Module. Though it might make sense to get all unsassigned request too (meaning call get_pend_reqs()) when being the primary. We need to ask this question to Ioannis. I've added it in the document.

Reply:
Will use `self.rep[self.id][PEND_REQS]` for now and add a `TODO` in the code to re-visit this when we know more.